reference-implementation/lib/import-maps.js

```
@@ -0,0 +1,117 @@
+'use strict';
+const { URL } = require('url');
+
+// https://fetch.spec.whatwg.org/#fetch-scheme
+const FETCH_SCHEMES = new Set(['http', 'https', 'ftp', 'about', 'blob', 'data', 'file', 'filesystem']);
+
+// Tentative, so better to centralize so we can change in one place as necessary (including tests).
+exports.BUILT_IN_MODULE_PREFIX = '@std/';
+
+exports.parseFromString = (input, baseURL) => {
+  const parsed = JSON.parse(input);
+
+  if (!isJSONObject(parsed)) {
+    throw new TypeError('Import map JSON must be an object.');
+  }
+
+  if ('imports' in parsed && !isJSONObject(parsed.imports)) {
+    throw new TypeError('Import map\'s imports value must be an object.');
+  }
+
+  if ('scopes' in parsed && !isJSONObject(parsed.scopes)) {
+    throw new TypeError('Import map\'s scopes value must be an object.');
+  }
+
+  if ('imports' in parsed) {
+    normalizeSpecifierMap(parsed.imports, baseURL);
+  }
+
+  const normalizedScopes = {};
+  if ('scopes' in parsed) {
+    for (const [scopeKey, specifierMap] of Object.entries(parsed.scopes)) {
+      if (!isJSONObject(specifierMap)) {
+        throw new TypeError(`The value for the "${scopeKey}" scope must be an object.`);
+      }
+
+      normalizeSpecifierMap(specifierMap, baseURL);
+
+      const scopeKeyURL = tryURLParse(scopeKey, baseURL);
+      if (scopeKeyURL === null) {
+        continue;
+      }
```

I'd still prefer to be strict and throw errors for invalid cases.